core/src/main/scala/kafka/log/LazyIndex.scala

```diff
@@ -0,0 +1,147 @@
+import java.util.concurrent.locks.ReentrantLock
+
+import LazyIndex._
+import kafka.utils.CoreUtils.inLock
+import kafka.utils.threadsafe
+import org.apache.kafka.common.utils.Utils
+
+/**
+  * A wrapper over an `AbstractIndex` instance that provides a mechanism to defer loading (i.e. memory mapping) the
+  * underlying index until it is accessed for the first time via the `get` method.
+  *
+  * This is an important optimization with regards to broker start-up and shutdown time if it has a large number of segments.
+  * It prevents illegal accesses to the underlying index after closing the index, which might otherwise lead to memory
+  * leaks due to recreation of underlying memory mapped object.
+  *
+  * Finally, this wrapper ensures that redundant disk accesses and memory mapped operations are avoided upon attempts to
+  * delete or rename the file that backs this index.
+  *
+  * @param loadIndex A function that takes a `File` pointing to an index and returns a loaded `AbstractIndex` instance.
+  */
+@threadsafe
+class LazyIndex[T <: AbstractIndex] private (@volatile private var indexWrapper: IndexWrapper, loadIndex: File => T) {
+  // A closed index does not allow accessing its indices to prevent side effects.
+  @volatile private var isClosed: Boolean = false
+  private val lock = new ReentrantLock()
+
+  def file: File = indexWrapper.file
+
+  def file_=(f: File): Unit = {
+    inLock(lock) {
+      indexWrapper.file = f
+    }
+  }
+
+  def get: T = {
+    if (isClosed)
+      throw new IllegalStateException(s"Attempt to access the closed Index (file=$file).")
+    indexWrapper match {
+      case indexValue: IndexValue[T] => indexValue.index
+      case _: IndexFile =>
+        inLock(lock) {
+          indexWrapper match {
+            case indexValue: IndexValue[T] => indexValue.index
+            case indexFile: IndexFile =>
+              val indexValue = new IndexValue(loadIndex(indexFile.file))
+              indexWrapper = indexValue
+              indexValue.index
+          }
+        }
+    }
+  }
+
+  /**
+   * Close this index file.
+   * Note: This will be a no-op if the index has already been closed.
+   */
+  def close(): Unit = {
+    if (!isClosed) {
+      inLock(lock) {
+        indexWrapper match {
+          case indexValue: IndexValue[T] => indexValue.index.close()
+          case _: IndexFile => // no-op
+        }
+      }
+      isClosed = true
+    }
+  }
+
+  /**
+   * Delete the index file that backs this index if exists.
+   * This method ensures that if the index file has already been closed, it will not be recreated as a side effect.
+   *
+   * @throws IOException if deletion fails due to an I/O error
+   * @return `true` if the file was deleted by this method; `false` if the file could not be deleted because it did
+   *         not exist
+   */
+  def deleteIfExists(): Boolean = {
+    if (isClosed)
+      Files.deleteIfExists(file.toPath)
+    else
+      get.deleteIfExists()
```

Comment:
Sorry for being late for the review. I have a comment that may also be related to #65.

There are three possible states of a lazy index:
1. Uninitialized (not loaded)
2. Loaded and not closed
3. Loaded and closed

The code here indicates that we will call get for both case 1 and 2 meaning that we will also load the index if the lazy index has not yet been loaded. If we want to be ultra lazy, do you think we should also avoid loading the index for case 1) like what we did in the close method.

Reply:
Despite its potential indirect help on overall resource usage, I feel that the proposed improvement is orthogonal to this patch -- i.e. this one is specifically about a cherry-pick that addresses `BufferOverflow` exceptions (and also to #65, whose main target is to improve the shutdown performance).
A call to this function is relevant in cases including (1) aborting a log segment cleanup into a single replacement segment, (2) aborting splitting a single log segment into multiple log segments, (3) deleting old segments after swapping them with new ones, (4) deleting corrupted and inconsistent segments during recovery, (5) splitting legacy log segments with offset overflow during segment loading, and (6) deleting old segments for retention. To keep this improvement proposal self-contained, I suggest exploring this proposal in its relevant PR. 
-- Please let me know what you think.

Reply:
I agree with you. We should not introduce any change other than the cherry pick in this PR. We can evaluate whether we do need to be ultra lazy for this method and open a separate PR if needed.